crates/arrow_sql_gen/src/clickhouse.rs

```diff
@@ -0,0 +1,546 @@
+    FailedToGetRowValue {
+        clickhouse_type: SqlType,
+        source: clickhouse_rs::errors::Error,
+    },
+
+    #[snafu(display("Failed to append a row value for {}: {}", clickhouse_type, source))]
+    FailedToAppendRowValue {
+        clickhouse_type: SqlType,
+        source: arrow::error::ArrowError,
+    },
+
+    #[snafu(display("No Arrow field found for index {index}"))]
+    NoArrowFieldForIndex { index: usize },
+
+    #[snafu(display("No column name for index: {index}"))]
+    NoColumnNameForIndex { index: usize },
+
+    #[snafu(display("Cannot represent BigDecimal as i128: {big_decimal}"))]
+    FailedToConvertBigDecimalToI128 { big_decimal: BigDecimal },
+
+    #[snafu(display("Failed to parse decimal string as BigInterger {}: {}", value, source))]
+    FailedToParseBigDecimalFromClickhouse {
+        value: String,
+        source: bigdecimal::ParseBigDecimalError,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+macro_rules! handle_primitive_type {
+    ($builder:expr, $type:expr, $builder_ty:ty, $value_ty:ty, $row:expr, $index:expr) => {{
+        let Some(builder) = $builder else {
+            return NoBuilderForIndexSnafu { index: $index }.fail();
+        };
+        let Some(builder) = builder.as_any_mut().downcast_mut::<$builder_ty>() else {
+            return FailedToDowncastBuilderSnafu {
+                clickhouse_type: format!("{:?}", $type),
+            }
+            .fail();
+        };
+        let v = $row
+            .get::<$value_ty, usize>($index)
+            .context(FailedToGetRowValueSnafu {
+                clickhouse_type: $type,
+            })?;
+
+        builder.append_value(v)
+    }};
+}
+
+macro_rules! handle_primitive_nullable_type {
+    ($builder:expr, $type:expr, $builder_ty:ty, $value_ty:ty, $row:expr, $index:expr) => {{
+        let Some(builder) = $builder else {
+            return NoBuilderForIndexSnafu { index: $index }.fail();
+        };
+        let Some(builder) = builder.as_any_mut().downcast_mut::<$builder_ty>() else {
+            return FailedToDowncastBuilderSnafu {
+                clickhouse_type: format!("{:?}", $type),
+            }
+            .fail();
+        };
+        let v = $row
+            .get::<Option<$value_ty>, usize>($index)
+            .context(FailedToGetRowValueSnafu {
+                clickhouse_type: $type,
+            })?;
+
+        match v {
+            Some(v) => builder.append_value(v),
+            None => builder.append_null(),
+        }
+    }};
+}
+
+/// Converts `Clickhouse` `Block` to an Arrow `RecordBatch`. Assumes that all rows have the same schema and
+/// sets the schema based on the `sql_type` returned for column.
+///
+/// # Errors
+///
+/// Returns an error if there is a failure in converting the rows to a `RecordBatch`.
+#[allow(clippy::too_many_lines)]
+pub fn block_to_arrow(block: &Block<Complex>) -> Result<RecordBatch> {
```

Comment:
This looks like a good candidate for a simple unit test that tests conversion of a Clickhouse Block to an Arrow RecordBatch

Reply:
Probably better to do it in a separate pr.